File: Java/109. Convert Sorted List to Binary Search Tree.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode sortedListToBST(ListNode head) {
        if (head == null) {
            return null;
        }

        ListNode mid = findMid(head);
        TreeNode root = new TreeNode(mid.val);

        if (mid == head) {
            return root;
        }
        root.left = sortedListToBST(head);
        root.right = sortedListToBST(mid.next);
        return root;
    }

    private ListNode findMid(ListNode root) {
        ListNode prev = null, mid = root, fast = root;
        while (fast != null && fast.next != null) {
            prev = mid;
            mid = mid.next;
            fast = fast.next.next;
        }
        if (prev != null) {
            prev.next = null;
        }
        return mid;
    }
}

File: Python3/109. Convert Sorted List to Binary Search Tree.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedListToBST(self, head: Optional[ListNode]) -> Optional[TreeNode]:
        if not head:
            return None

        # return mid
        def findMid(root: Optional[ListNode]) -> Optional[ListNode]:
            prev = None
            mid = fast = root
            while fast and fast.next:
                prev = mid
                mid = mid.next
                fast = fast.next.next

            if prev:
                prev.next = None
            
            return mid

        def helper(root: Optional[ListNode]) -> Optional[TreeNode]:
            if not root:
                return None
                
            mid = findMid(root)
            new_root = TreeNode(mid.val)

            # only one node left
            if root == mid:
                return new_root

            new_root.left = helper(root)
            new_root.right = helper(mid.next)
            return new_root

        return helper(head)

